backend/db/utils.py

```diff
@@ -130,21 +182,107 @@ def sql_query(
+    :return: dictionary with keys as the primary key and values as the row contents"""
+    key_vals = {key: [str(obj[key]) for obj in objs] for key in keys}
+    results: List[RowMapping] = sql_query(
+        con, f"SELECT * FROM {table} WHERE {' AND '.join([f'{k} {sql_in(v, True)}' for k, v in key_vals.items()])};",
+        return_with_keys=True)
+    return [dict(x) for x in results]
+
+
+def get_concept_set_members_rows(
+    con: Connection, codeset_id__concept_id__pairs: List[Tuple[int, int]]
+) -> List[Dict]:
+    """Get rows from concept_set_members given IDs
+    todo: too specific to be in utils.py? move?"""
+    objs = [{'codeset_id': pair[0], 'concept_id': pair[1]} for pair in codeset_id__concept_id__pairs]
+    return get_objs_by_composite_key(con, 'concept_set_members',['codeset_id', 'concept_id'], objs)
+
+
+def get_cset_members_items_rows(
+    con: Connection, codeset_id__concept_id__pairs: List[Tuple[int, int]]
+) -> List[Dict]:
+    """Get rows from cset_members_items given IDs
+    todo: too specific to be in utils.py? move?"""
+    objs = [{'codeset_id': pair[0], 'concept_id': pair[1]} for pair in codeset_id__concept_id__pairs]
+    return get_objs_by_composite_key(con, 'cset_members_items', ['codeset_id', 'concept_id'], objs)
+
+
+def insert_from_dicts(con: Connection, table: str, rows: List[Dict], skip_if_already_exists=True):
+    """Insert rows into table from a list of dictionaries"""
+    pk: str = pkey(table)
     if skip_if_already_exists:
-        pk = pkey(table)
+        if pk and isinstance(pk, str):  # normal, single primary key
+            already_in_db: List[Dict] = get_objs_by_id(con, table, pk, [row[pk] for row in rows])
+            already_in_db_ids = [row[pk] for row in already_in_db]
+            rows = [row for row in rows if row[pk] not in already_in_db_ids]
+        elif pk and isinstance(pk, list):  # composite key
+            already_in_db: List[Dict] = get_objs_by_composite_key(con, table, pk, rows)
+            already_in_db_ids = [[row[pk_n] for pk_n in pk] for row in already_in_db]
+            rows = [row for row in rows if [row[pk_n] for pk_n in pk] not in already_in_db_ids]
+
+    if rows:
+        # Fix possible jaggedneess / missing fields in rows
+        fields = []
+        for row in rows:
+            fields.extend(row.keys())
+            for k in row.keys():
+                fields.append(k)
+        fields = set(fields)
+        rows = [{field: row.get(field, None) for field in fields} for row in rows]
+        key_vals = {f'{k}{i}': v for i, d in enumerate(rows) for k, v in d.items()}
+        values = ', '.join([f"({', '.join([':' + str(k) + str(i) for k in d.keys()])})" for i, d in enumerate(rows)])
+        statement = f"""INSERT INTO {table} ({', '.join([f'"{x}"' for x in rows[0].keys()])}) VALUES {values}"""
+        run_sql(con, statement, key_vals)
+
+
+def insert_from_dict(con: Connection, table: str, d: Union[Dict, List[Dict]], skip_if_already_exists=True):
```

Comment:
`insert_from_dict` `skip_if_already_exists` check: Added logic for if table has composite key, e.g. `concept_set_members`